docs/essentials/collections/threadsafe/how toAddandRemoveItems.md

```diff
@@ -1 +1,285 @@
+                {
+                    //The data was not updated. Log error, throw exception, etc.
+                    Console.WriteLine("Could not update {0}", retrievedValue.Name);
+                }
+            }
+            else
+            {
+                // Add the new key and value. Here we call a method to retrieve
+                // the data. Another option is to add a default value here and 
+                // update with real data later on some other thread.
+                CityInfo newValue = GetDataForCity(searchKey);
+                if( cities.TryAdd(searchKey, newValue))
+                {
+                    // use the data
+                    Console.Write("Most recent high temperatures for {0} are: ", newValue.Name);
+                    int[] temps = newValue.RecentHighTemperatures;
+                    foreach (var temp in temps) Console.Write("{0}, ", temp);
+                }
+                else
+                    Console.WriteLine("Unable to add data for {0}", searchKey);
+            }
+        }
+
+        //Assume this method knows how to find long/lat/temp info for any specified city.
+        static CityInfo GetDataForCity(string name)
+        {
+            // Real implementation left as exercise for the reader.
+            if (String.CompareOrdinal(name, "Caracas") == 0)
+                return new CityInfo() { Name = "Caracas", 
+                                        Longitude = 10.5M, 
+                                        Latitude = -66.916667M,
+                                        RecentHighTemperatures = new int[] { 91, 89, 91, 91, 87, 90, 91 } };
+            else if (String.CompareOrdinal(name, "Buenos Aires") == 0)
+                return new CityInfo() { Name = "Buenos Aires", 
+                                        Longitude = -34.61M, 
+                                        Latitude = -58.369997M, 
+                                        RecentHighTemperatures = new int[] { 80, 86, 89, 91, 84, 86, 88 } };
+            else
+                throw new ArgumentException("Cannot find any data for {0}", name);
+        }
+    }
+}
+
+```
+[ConcurrentDictionary&lt;TKey, TValue&gt;](http://dotnet.github.io/api/System.Collections.Concurrent.ConcurrentDictionary%602.html) is designed for multithreaded scenarios. You do not have to use locks in your code to add or remove items from the collection. However, it is always possible for one thread to retrieve a value, and another thread to immediately update the collection by giving the same key a new value.
+
+Also, although all methods of **ConcurrentDictionary&lt;TKey, TValue&gt;** are thread-safe, not all methods are atomic, specifically **GetOrAdd** and **AddOrUpdate**. The user delegate that is passed to these methods is invoked outside of the dictionary's internal lock. (This is done to prevent unknown code from blocking all threads.) Therefore it is possible for this sequence of events to occur:
```

Comment:
Here the article is assuming that the reader understands that there are two delegates at play here.  Is that implementation detail documented somewhere so that this can be modified to call that out?